### ISSUE TITLE
Calling Function or eval should be eliminated as a security measure and would cause the application execution to be halted if the document CSP header has script-src 'self';

### ISSUE THREAD

**Issue by bakrkawkji-trustly** [webpack/webpack#5627]
https://github.com/webpack/webpack/blob/fb8afe71385734cfd65f47949117306a91f20753/buildin/global.js#L10

**Comment by bakrkawkji-trustly** [comment]
@will-stone yeah this issue still exists, my current local solution for it is to run the following bash command to replace it with undefined before deploying my code:

"find /assets/ -type f -name \"*.js\" | xargs sed -i '' 's/([0,1], eval)(\"this\")/undefined/g'"

**Comment by balmas** [comment]
Any news on a fix for this?

**Comment by hotrush** [comment]
Any news?

**Comment by evilebottnawi** [comment]
No, we need `eval` here. Solution:
1. https://csp-evaluator.withgoogle.com/
```
script-src 'strict-dynamic' 'nonce-rAnd0m123' 'unsafe-inline' http: https:;
object-src 'none';
base-uri 'none';
report-uri https://csp.example.com;
```

**Comment by bakrkawkji-trustly** [comment]
@evilebottnawi can you explain why eval is needed please?

**Comment by evilebottnawi** [comment]
@bakrkawkji-trustly return global context

**Comment by bakrkawkji-trustly** [comment]
@evilebottnawi let me rephrase my question:
In what case `(1,eval)("this")` will be executed? isn't `Function("return this")()` is healthy enough way to return the global context without using eval? if `Function("return this")()` fails, it's more likely there's something wrong with the browser, so why trying to return the global context using `(1,eval)("this")`? I find this as an overkill way to obtain the global context, not quite justified as a need but as a "do whatever it takes" approach

**Comment by graingert** [comment]
@bakrkawkji-trustly Function(str) is eval.

**Comment by akanshgulati** [comment]
Can we have some provision to avoid evals in build totally. I tried same code with browserify and vueify which was earlier build with webpack and vue-loader and I found exactly 0 issues of eval or function constructor.

**Comment by torarnv** [comment]
The Vue documentation describes how "On the other hand, the runtime-only build is fully CSP-compliant. When using the runtime-only build with Webpack + vue-loader or Browserify + vueify, your templates will be precompiled into render functions which work perfectly in CSP environments."

https://vuejs.org/v2/guide/installation.html#CSP-environments

So coming from Vue, I'd expect to be able to use Webpack to bundle up mye Vue app as something that can be run in a Chrome extension.

**Comment by volokolamskspb** [comment]
change your webpack.config to this:
```js 
const path = require('path');
module.exports = {
  devtool: "source-map",
  entry: ['babel-polyfill', path.resolve(__dirname, './yourjs.js')],
  mode: 'production',
  output: {
    path: __dirname+'/draft.iml.ru/static/main/js/',
    filename: 'yout.js'
  }
};
```
had to add devtools option.

**Comment by jhuesos** [comment]
We also have CSP that disallows eval()/Function(str) for security reasons and I have not figured out any way to produce production code without eval(). And the `'nonce-rAnd0m123'` is not an option because the CSP headers are being set by our authentication layer so we can't just add nonce- for all webpack files.

The previous proposed solutions did not solve the problem for us.

**Comment by jkobylec** [comment]
@jhuesos, I suspect this is because your webpack config does not explicitly set `global` to false on the `node` object, so it injects the global.js polyfill which uses` eval`/`new Function`.

For instance, I was also having this problem with the default webpack config from the 'webpack' template for Vue CLI. So I modified the below block in build\webpack.base.conf.js:

```JavaScript
node: {
    // prevent webpack from injecting useless setImmediate polyfill because Vue
    // source contains it (although only uses it if it's native).
    setImmediate: false,
    // prevent webpack from injecting mocks to Node native modules
    // that does not make sense for the client
    dgram: 'empty',
    fs: 'empty',
    net: 'empty',
    tls: 'empty',
    child_process: 'empty',
    // prevent webpack from injecting eval / new Function through global polyfill
    global: false
}
```

For others paying attention to this thread, I am curious whether this could have performance implications as well? Is it still true that the presence of `eval` or `new Function` at all effectively causes other code on the page to lose out on a lot of JS engine optimizations? Or is this one of those things that was more true when Crockford said so, but is a little more nuanced now? I'm not seeing much conclusive analysis out there.

I am also wondering if there might be some value to allowing an option in configuration of injecting a more "naïve" shim for the global object which omits the `eval`/`new Function` avenue but still tries the other things? I realize it won't cover as many cases, but seems like it might be preferable under some circumstances.

**Comment by bakrkawkji-trustly** [comment]
@jkobylec Finally a solution that solves this issue! thanks!

For anyone else who's still looking for a solution to remove:
` g = g || Function("return this")() || (1,eval)("this"); `

Use:
```
node: {
global: false
}
```

**Comment by xitter** [comment]
While above solutions for turning of global works, there are few libraries(e.x. [babel-es6-polyfill](https://github.com/JasonBerry/babel-es6-polyfill/blob/master/polyfill.js#L3)) which refer to window variable via global and thus webpack bundling fails. So to be extra cautious, define a plugin for global as well.

```
{
  node: {
     global: false
  },
  plugins: new webpack.DefinePlugin({
     global: 'window'		// Placeholder for global used in any node_modules
  })
}
```

**Comment by yvele** [comment]
From what I understood, script loader [NEEDS](https://github.com/webpack-contrib/script-loader/issues/6#issuecomment-212367261) to use `eval` and then is incompatible with a strict `script-src` CSP required, and needs `unsafe-inline` which is not acceptable for any sensible application.

So is it possible, at least, to add a warning about that issue in the readme?

**Comment by depoulo** [comment]
> From what I understood, script loader NEEDS to use eval

This is no longer true:
- https://mathiasbynens.be/notes/globalthis
- https://github.com/ungap/global-this

**Comment by tcstory** [comment]
> @jkobylec Finally a solution that solves this issue! thanks!
> 
> For anyone else who's still looking for a solution to remove:
> `g = g || Function("return this")() || (1,eval)("this");`
> 
> Use:
> 
> ```
> node: {
> global: false
> }
> ```

I solved the problem by only setting devtool to "cheap-module-source-map"

**Comment by douglasb99** [comment]
> No, we need `eval` here. Solution is `strict-dynamic` directive:
> 
> 1. https://csp-evaluator.withgoogle.com/
> 
> ```
> script-src 'strict-dynamic' 'nonce-rAnd0m123' 'unsafe-inline' http: https:;
> object-src 'none';
> base-uri 'none';
> report-uri https://csp.example.com;
> ```

this does not work when you are trying to rub your code in various environments which require strict CSP eg chrome extensions

**Comment by TazmanianDI** [comment]
We are hitting this as well and none of the solutions above seem to be working for us. I've narrowed our specific problem down to our use of code splitting and customizing the public path. We make use of code splitting though the import `webpackChunkName` syntax and we also use `__webpack_public_path__` to dynamically set the path since we host our files on a CDN and since we use a form of versioning, the path is dynamic. It seems that use of either of these things results in this `webpack/runtime/global` code with `new Function` being emitted.

The following code all by itself does this: 

```
declare let __webpack_public_path__: string;
__webpack_public_path__ = `/versions/`;

function doSomething() {
  return 0;
}

export { doSomething };
```

We've tried the `node: { global: false }` change as well as the various values for `devtool`. If I eliminate the use of `__webpack_public_path__` but still have `webpackChunkName` then the code is still emitted. If we disable chunking by using `LimitChunkCountPlugin` then the code is *not* emitted.
